src/include/txn.i

```diff
@@ -538,34 +543,77 @@ __wt_txn_visible(
 #endif
 }
 
+/*
+ * __wt_txn_upd_visible_type --
+ *      Visible type of given update for the current transaction.
+ */
+static inline WT_VISIBLE_TYPE
+__wt_txn_upd_visible_type(WT_SESSION_IMPL *session, WT_UPDATE *upd)
+{
+	uint8_t upd_state;
+	bool upd_visible;
+
+	for (;;) {
+		if ((upd_state = upd->state) == WT_UPDATE_STATE_LOCKED) {
+			/* Commit is in progress, yield and try again. */
+			__wt_yield();
+			continue;
+		}
+
+		upd_visible = __wt_txn_visible(
+		    session, upd->txnid, WT_TIMESTAMP_NULL(&upd->timestamp));
+
+		/*
+		 * The visibility check is only valid if the update does not
+		 * change state.  If the state does change, recheck visibility.
+		 */
+		if (upd->state != upd_state)
+			continue;
+
+		if (!upd_visible)
+			return (WT_VISIBLE_FALSE);
+
+		if (upd_state == WT_UPDATE_STATE_PREPARED)
+			return (F_ISSET(&session->txn, WT_TXN_IGNORE_PREPARE) ?
+			    WT_VISIBLE_FALSE : WT_VISIBLE_PREPARE);
+
+		return (WT_VISIBLE_TRUE);
+	}
+}
```

Comment:
I'd recommend a `/* NOTREACHED */` after the for loop, it makes some static analyzers happier.